src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1391,18 +1391,46 @@ abstract class RefChecks extends Transform {
       }
     }
 
-    private def checkAnnotations(tpes: List[Type], tree: Tree) = tpes foreach { tp =>
-      checkTypeRef(tp, tree, skipBounds = false)
-      checkTypeRefBounds(tp, tree)
-    }
-    private def doTypeTraversal(tree: Tree)(f: Type => Unit) = if (!inPattern) tree.tpe foreach f
-
     private def applyRefchecksToAnnotations(tree: Tree): Unit = {
-      def applyChecks(annots: List[AnnotationInfo]) = {
-        checkAnnotations(annots map (_.atp), tree)
-        transformTrees(annots flatMap (_.args))
+      def applyChecks(annots: List[AnnotationInfo]): List[AnnotationInfo] = {
+        annots.foreach { ann =>
+          checkTypeRef(ann.tpe, tree, skipBounds = false)
+          checkTypeRefBounds(ann.tpe, tree)
+        }
+
+        annots
+          .map(_.transformArgs(transformTrees))
+          .groupBy(_.symbol)
+          .flatMap((groupRepeatableAnnotations _).tupled)
+          .toList
       }
 
+      // assumes non-empty `anns`
+      def groupRepeatableAnnotations(sym: Symbol, anns: List[AnnotationInfo]): List[AnnotationInfo] =
+        if (!sym.isJavaDefined) anns else anns match {
```

Comment:
@hrhino this is slightly different than in 2.12.x (https://github.com/scala/scala/pull/6846/files#diff-e2cfc09e4d7c753232666c6e31bd1926R1424)

Reply:
yep, thanks for adapting it!